page_iterator.go

```diff
@@ -0,0 +1,195 @@
+package msgraphgocore
+
+import (
+	"log"
+	"net/url"
+	"reflect"
+	"unsafe"
+
+	abstractions "github.com/microsoft/kiota/abstractions/go"
+	"github.com/microsoft/kiota/abstractions/go/serialization"
+	jsonserialization "github.com/microsoft/kiota/serialization/go/json"
+)
+
+type Page interface {
+	getValue() []interface{}
+	getNextLink() *string
+}
+
+// PageIterator represents an iterator object that can be used to get subsequent pages of a collection.
+type PageIterator struct {
+	currentPage     Page
+	reqAdapter      GraphRequestAdapterBase
+	pauseIndex      int
+	constructorFunc ParsableConstructor
+	headers         map[string]string
+}
+
+type ParsableConstructor func() serialization.Parsable
+
+type PageResult struct {
+	nextLink *string
+	value    []interface{}
+}
+
+func (p *PageResult) getValue() []interface{} {
+	if p == nil {
+		return nil
+	}
+
+	return p.value
+}
+
+func (p *PageResult) getNextLink() *string {
+	if p == nil {
+		return nil
+	}
+
+	return p.nextLink
+}
+
+// NewpageIterator creates an iterator instance
+//
+// It has three parameters. res is the graph response from the initial request and represents the first page.
+// reqAdapter is used for getting the next page and constructorFunc is used for serializing next page's response to the specified type.
+func NewPageIterator(res interface{}, reqAdapter GraphRequestAdapterBase, constructorFunc ParsableConstructor) *PageIterator {
+	abstractions.RegisterDefaultSerializer(func() serialization.SerializationWriterFactory {
```

Comment:
please remove those initialization from here, in real life people will have instantiated a new client and made a request, thus initializing this prior to calling this ctor. If you need it for unit testing, please have this code as part of your Arrange section